ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/recovery/arjunacore/SubordinateAtomicActionRecoveryModule.java

```diff
@@ -0,0 +1,99 @@
+ * This program is distributed in the hope that it will be useful, but WITHOUT A
+ * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+ * PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more details.
+ * You should have received a copy of the GNU Lesser General Public License,
+ * v.2.1 along with this distribution; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
+ * MA  02110-1301, USA.
+ *
+ * (C) 2016,
+ * @author JBoss Inc.
+ */
+package com.arjuna.ats.internal.jta.recovery.arjunacore;
+
+import com.arjuna.ats.arjuna.common.Uid;
+import com.arjuna.ats.arjuna.exceptions.ObjectStoreException;
+import com.arjuna.ats.arjuna.objectstore.RecoveryStore;
+import com.arjuna.ats.arjuna.objectstore.StoreManager;
+import com.arjuna.ats.arjuna.recovery.RecoveryModule;
+import com.arjuna.ats.arjuna.state.InputObjectState;
+import com.arjuna.ats.internal.arjuna.common.UidHelper;
+import com.arjuna.ats.internal.jta.transaction.arjunacore.jca.SubordinationManager;
+import com.arjuna.ats.internal.jta.transaction.arjunacore.subordinate.jca.SubordinateAtomicAction;
+import com.arjuna.ats.jta.logging.jtaLogger;
+
+import javax.transaction.xa.XAException;
+import java.io.IOException;
+
+
+/**
+ * This class is purely used by the recovery system to load the transactions into memory so we can be sure
+ * that orphan detection can find the TransactionImple if the EIS has not called XATerminator::recover yet
+ *
+ * At the time of writing this module is utilized by the SubordinationManagerXAResourceOrphanFilter to ensure that
+ * it can check with the SubordinationManager if the transaction is in flight during orphan detection.
+ */
+public class SubordinateAtomicActionRecoveryModule implements RecoveryModule {
+
+    private boolean recoverySucceeded;
+
+    @Override
+    public void periodicWorkFirstPass() {
+        /*
+         * Requires going through the objectstore for the states of imported
+         * transactions - this is just to make sure the server control are loaded into memory.
+         *
+         * The EIS will call XATerminator::recover() for actual crash recovery
+         */
+
+        RecoveryStore recoveryStore = StoreManager.getRecoveryStore();
+        InputObjectState states = new InputObjectState();
+
+        // only look in the JCA section of the object store
+        Uid uid = null;
+
+        try {
+            if (recoveryStore.allObjUids(SubordinateAtomicAction.getType(), states)
+                    && (states.notempty())) {
+                while (true) {
+                    UidHelper.unpackFrom(states);
+                    if (uid.notEquals(Uid.nullUid())) {
+                        SubordinationManager
+                                .getTransactionImporter().recoverTransaction(
+                                uid);
+                    } else {
+                        break;
```

Comment:
Here you set recoverySucceeded to true even though nothing was recovered.

Reply:
I have changed the name of this and commented where it is used to try to make that clearer

Reply:
okay